### DISCUSSION TITLE
decorator problem in `mass` when re-routing to non-normalized function

### DISCUSSION THREAD

**Discussion by NimaSarajpoor** [TDAmeritrade/stumpy#530] [General]
I recently noticed that there is a need to provide the name of arguments in `mass` function if I want to set `normalize=False`. 

In other words, I cannot simply do:

```
T, sliding_mu, sliding_std = core.preprocess(T, m)
dist_profile = core.mass(Q, T, sliding_mu, sliding_std, normalize=False)
```
This gives me the following error:
```
TypeError                                 Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_8412/711003992.py in <module>
      1 T, sliding_mu, sliding_std = core.preprocess(T, m)
      2 Q = T[0:0+m]
----> 3 dist_profile = core.mass(Q, T, sliding_mu, sliding_std, normalize=False)

~\anaconda3\lib\site-packages\stumpy\core.py in inner_wrapper(*args, **kwargs)
    136                             else:
    137                                 kwargs[v] = kwargs.pop(k)
--> 138                 return non_norm(*args, **kwargs)
    139 
    140         return inner_wrapper

~\anaconda3\lib\site-packages\stumpy\core.py in mass_absolute(Q, T, T_subseq_isfinite, T_squared)
   1102             T_squared = np.sum(rolling_window(T * T, m), axis=-1)
   1103         distance_profile[:] = _mass_absolute(Q_squared, T_squared, QT)
-> 1104         distance_profile[~T_subseq_isfinite] = np.inf
   1105 
   1106     return distance_profile

TypeError: ufunc 'invert' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

And, instead I should do:
```
T, sliding_mu, sliding_std = core.preprocess(T, m)
dist_profile = core.mass(Q, T,  M_T=sliding_mu, Σ_T=sliding_std, normalize=False)
```
Is that by design? because, as a user, it is hard for me to type `Σ_T` and I have to copy it. 

(NOTE: I know `M_T` and `Σ_T` are not used in `core._mass_absolute`. I have no problem with using `mass_absolute`. I merely wanted to let you know of such behavior)

**Comment by NimaSarajpoor** [comment]
@seanlaw 
Btw, I took a look at the docstring of ` non_normalized` and, according to the example provided, it seems we should provide the name of arguments. However, I was wondering if there is a way round this matter.

**Reply by seanlaw** [reply]
So, `normalize=False` basically re-routes to `core.mass_absolute`, which means that you should really be doing:

```
T, T_subseq_isfinite = core.preprocess_non_normalized(T, m)
dist_profile = core.mass(Q, T, normalize=False)
```

This should work and this should produce the same results as:

```
T, T_subseq_isfinite = core.preprocess_non_normalized(T, m)
dist_profile = core.mass_absolute(Q, T)
```

**Reply by NimaSarajpoor** [reply]
Yes, I noticed that when I went to do some investigation after  getting that error. 



> dist_profile = core.mass(Q, T, normalize=False)

Exactly. So, better not to provide the other two arguments (i.e. `M_T` and `Σ_T`) in the first place if we want to set `normalize=False`.  

Thanks.

**Reply by seanlaw** [reply]
Notice also that we use `core.preprocess_non_normalized` when we use `normalize=False` and not `core.preprocess`

**Reply by NimaSarajpoor** [reply]
I checked out the changes. Got it.  Thanks for the effort!!

**Reply by seanlaw** [reply]
So, let's say you have `1000` queries that you want to compute the non-normalized distance profiles for:

```
Qs = np.random.uniform(-10, 10, size=10000).reshape(1000, 10)
m = Qs.shape[1]
T = np.random.rand(1000)
```

You could do:
```
T, T_subseq_isfinite = core.preprocess_non_normalized(T, m)
dist_profiles = []
for i, Q in enumerate(Qs):
    dist_profiles[i] = core.mass(Q, T, normalize=False)
    # Equivalent: dist_profiles[i] = core.mass_absolute(Q, T)
```

However, internally, `core.mass_absolute` will compute `T_subseq_isfinite` and `T_squared` for every iteration and this can be wasteful and time consuming. So, since `T_subseq_isfinite` and `T_squared` doesn't change in each iteration, instead, you can tell `core.mass_absolute` (or `core.mass`) to use a pre-computed `T_subseq_isfinite` and `T_squared`:

```
T, T_subseq_isfinite = core.preprocess_non_normalized(T, m)
T_squared = np.sum(core.rolling_window(T * T, m), axis=-1)
dist_profiles = []
for i, Q in enumerate(Qs):
    dist_profiles[i] = core.mass(Q, T, M_T=T_subseq_isfinite, Σ_T=T_squared, normalize=False)
    # Equivalent: dist_profiles[i] = core.mass_absolute(Q, T, T_subseq_isfinite, T_squared)
```

Essentially, in the background the key `M_T` is being replaced by `T_subeq_isfinite` and the key `Σ_T` is replaced by `T_squared` (but the value stays the same). It's not the best thing and I don't encourage the average user to do this and nor should 99.99999% need to know about this detail. If you really intend to do this then you should just use `core.mass_absolute` instead or `core.mass` without specifying `M_T` and `Σ_T`. I believe that this will still work too:

```
dist_profiles[i] = core.mass(Q, T, T_subseq_isfinite=T_subseq_isfinite, T_squared=T_squared, normalize=False)
```
And it will produce the same result. Again, all of this is for convenience and the non_normalized decorator is providing the "intelligent" flexibility.

Let me know if that makes sense.

**Reply by NimaSarajpoor** [reply]
Yes, that totally makes sense! Thanks for the explanation.

> If you really intend to do this then you should just use `core.mass_absolute` instead or `core.mass` without specifying `M_T` and `Σ_T`.

Correct as the other approach might crease confusion later.

**Reply by seanlaw** [reply]
The key reason for doing this is that I didn't want two public functions (z-norm and non-norm) for everything in our public API. This design approach (that I came up with) lets us manage things easily in the background and prevents our code base from being littered with tons and tons of:

```
if normalize:
    # run this giant chunk of code
else:
    # run this other very similar but slightly different giant chunk of code
```

Instead the z-norm and non-norm code exists cleanly in separate files and their functions are tested separately without needing to test for branching of the above if/else. It requires a bit of work to remember to make changes in both places but it's a small price to pay for clean code.

**Reply by NimaSarajpoor** [reply]
Yes, it is a VERY good design indeed!!

**Reply by seanlaw** [reply]
@NimaSarajpoor I can't remember but did we already fix this or resolve this in the code? If so, we should probably put a reference here

**Reply by NimaSarajpoor** [reply]
> but did we already fix this or resolve this in the code?

tbh, I cannot remember the very first reason that causes the issue in my case. But, I think that you changed the decorator while we were discussing this issue. So, I think `replace` was something else and then you changed it to `replace={"M_T": "T_subseq_isfinite", "Σ_T": None}` (I am not sure though)

> Essentially, in the background the key `M_T` is being replaced by `T_subeq_isfinite` and the key `Σ_T` is replaced by `T_squared` (but the value stays the same). It's not the best thing and I don't encourage the average user to do this and nor should 99.99999% need to know about this detail.

The problem is that the user should know that they MUST replace `M_T` with `T_subeq_isfinite` if they want to use `normalize=False`

**Reply by NimaSarajpoor** [reply]
@seanlaw 
I took a look at the top of this discussion again. It seems we didn't have this problem before:
```
T, sliding_mu, sliding_std = core.preprocess(T, m)
dist_profile = core.mass(Q, T,  M_T=sliding_mu, Σ_T=sliding_std, normalize=False) # I think this was working before

dist_profile = core.mass(Q, T,  sliding_mu, sliding_std, normalize=False) # I think this was NOT working before
```

----

So, you may want to do:
```
replace={"M_T": None, "Σ_T": None}
```
But then, we need to explain in the docstring that the user needs to pass `M_T` and `Σ_T` as keyword argument if they want  to do `normalize=False`. 

What do you think? 

~Personally, I think the aforementioned approach is better than `dist_profiles[i] = core.mass(Q, T, M_T=T_subseq_isfinite, Σ_T=T_squared, normalize=False)` (which is what one needs to do in the current version)~

**Reply by seanlaw** [reply]
> The problem is that the user should know that they MUST replace M_T with T_subeq_isfinite if they want to use normalize=False

So, the way that the decorator has been designed, it's actually "better" to not replace but to simply do:

```
_, T_subseq_isfinite = core.preprocess_non_normalized(dimensions, m)
dist_profile = core.mass(Q, T,  T_subseq_isfinite=T_subseq_isfinite, normalize=False)
```

When `normalize=False`, `T_subseq_isfinite` is recognized. The "replace" (detects whether `M_T` is present and then changes the `key` to `T_subseq_isfinite`) as a convenience but, realistically, we should avoid writing our code as:

```
_, T_subseq_isfinite = core.preprocess_non_normalized(dimensions, m)
dist_profile = core.mass(Q, T,  M_T=T_subseq_isfinite, normalize=False)
```

This will "work" but is super confusing.

> But then, we need to explain in the docstring that the user needs to pass M_T and Σ_T as keyword argument if they want to do normalize=False.

No, instead, we should explain in the docstring (in "normalize") that the user may pass `T_subseq_isfinite` as a keyword argument when `normalize=False`, though, it is optional.

**Reply by NimaSarajpoor** [reply]
> No, instead, we should explain in the docstring (in "normalize") that the user may pass `T_subseq_isfinite` as a keyword argument when `normalize=False`, though, it is optional.

This is definitely better!

**Reply by seanlaw** [reply]
Note that when `normalize=True`, passing `T_subseq_isfinite` in as a parameter would/should cause `core.mass` to raise an error

**Reply by seanlaw** [reply]
Added new issue #637 to track this

**Reply by seanlaw** [reply]
I'm not trying to pat myself on the back here as I couldn't remember "why" I had designed the decorator this way. However, this repeated issue actually helps me recall "why" and it seems to be working as intended and only requires adding a note to our docstring(s) without changing any code to accomodate. 😄

**Reply by NimaSarajpoor** [reply]
> I'm not trying to pat myself on the back here as I couldn't remember "why" I had designed the decorator this way

:D

> However, this repeated issue actually helps me recall "why" and it seems to be working as intended and only requires adding a note to our docstring(s) without changing any code to accomodate.

And, again, we can see the importance of having a good docstring. I am now  appreciating the great effort that has been put into making good docstrings in STUMPY. I recently started contributing to tslearn package (`improve KNN by using dtw lb_keogh` and I thought it can prepare me for SWAMP paper). The problem is that its docstring, in my opinion, is not good! I had to follow codes and keep them in mind to understand the parameters! So, I think writing a good docstring is VERY important! Thanks for your effort!

**Reply by seanlaw** [reply]
Yeah, the most important thing really is documentation AND unit testing. IMHO, it really doesn't make sense to focus on developing new features when the docs and tests are bad. This is my personal bias. Unfortunately, building new features is wayyy more fun than writing docs and tests and this latter part can be a thankless job.
